File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.c
/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     struct Node *next;
 *     struct Node *random;
 * };
 */
int listlen(struct Node *l){
    return (l == NULL) ? 0 : 1+listlen(l->next);
}

struct Node* copyRandomList(struct Node* head) {
    if (head == NULL)
        return NULL;
    int len = listlen(head);
    struct Node *random[len];
    int j = 0;

	struct Node* ans = malloc(sizeof(*ans));
    struct Node* ret = ans;
    struct Node* node = head;
    while (node != NULL){
        ans->val = node->val;
        random[j++] = node->random;
        node = node->next;
        if (node != NULL){
            ans->next = malloc(sizeof(*ans));
            ans = ans->next;
        } else {
            ans->next = NULL;
        }
    }
    j = 0;
    node = ret;
    for (int i = 0; i<len; ++i){
        struct Node *help = ret;
        struct Node *help2 = head;
        while (help2 != random[j]){
            help = help->next;
            help2 = help2->next;
        }
        node->random = random[j++] != NULL ? help : NULL;
        node = node->next;
    }

    return ret;
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x, next=None, random=None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: Node
        :rtype: Node
        """
        if not head:
            return None
        pending_randoms = {}
        i = 0
        new_head = Node(head.val)
        og_list = head
        runner = new_head
        while head:
            if head.random != None:
                pending_randoms.setdefault(head.random, []).append(runner)
            if head.next:
                new_node = Node(head.next.val)
                runner.next = new_node
                runner = runner.next
            head = head.next
        
        runner = new_head
        while og_list:
            if og_list in pending_randoms:
                for r in pending_randoms[og_list]:
                    r.random = runner
            runner = runner.next
            og_list = og_list.next

        return new_head

            
        